crates/sncast/src/helpers/fee.rs

```diff
@@ -106,29 +113,36 @@ impl FeeArgs {
                         ),
                         max_gas_unit_price: Some(max_gas_unit_price.try_into_()?),
                     },
-                    (Some(max_fee), Some(max_gas), None) => FeeSettings::Strk {
-                        max_gas: Some(max_gas.try_into_()?),
-                        max_gas_unit_price: Some(
-                            max_fee
-                                .floor_div(&NonZeroFelt::from_felt_unchecked(max_gas))
-                                .try_into_()?,
-                        ),
-                    },
+                    (Some(max_fee), Some(max_gas), None) => {
+                        let max_gas_unit_price =
+                            max_fee.floor_div(&NonZeroFelt::from_felt_unchecked(max_gas));
+                        if max_gas_unit_price == Felt::ZERO {
```

Comment:
The unchecked word should signal that it is not something we should use, as handling it incorrectly can lead to undefined behavior. In Rust, unchecked constructs like this are typically marked as `unsafe`. For example, see [`std::num::NonZero`](https://github.com/rust-lang/rust/blob/3378a5e084fa51c2e7f5c3f53b87e8ad89f1c517/library/core/src/num/nonzero.rs#L397). I’m not sure why Jonathan didn’t make it `unsafe`.

Previously, this functionality was clearly incorrect because we didn’t check whether it was zero. Now, we do check it, but the check happens *after* the operation. 

The sane approach would look something like this:

```rust
&NonZeroFelt::try_from(max_gas_unit_price)
    .context("--max-gas calculated from --max-fee should be greater than 0. Please increase --max-fee")?
```

Additionally, please remove the other places where `from_felt_unchecked` is used.

Reply:
Also (probably) the best way to handle this situation is to create our own wrapper type with the `FromStr` trait, which disallows the number `0`. This way, `clap`, while parsing the command, would immediately return an error:

```rs
struct NonZeroFelt(Felt);

impl FromStr for NonZeroFelt {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s.parse::<Felt>() {
            Ok(felt) => {
                if felt == Felt::ZERO {
                    Err(String::from("value should be greater than 0"))
                } else {
                    Ok(NonZeroFelt(felt))
                }
            }
            Err(_) => Err(String::from("value should be a number")),
        }
    }
}
```

Reply:
I’m sorry if I seem mean—I don’t intend to be. But together, we can forge a better place. 
![a1d2b4446d5bbf9659a5279470acae9c](https://github.com/user-attachments/assets/cc23d177-9f55-4184-9e03-7943efb5f9c3)

Reply:
> I’m not sure why Jonathan didn’t make it unsafe.

Because it is not how `unsafe` is designed to be used. It is not for invariant correctness but for memory correctness.
Rust's `NonZero` types have the `unchecked` method marked as `unsafe` because it can violate memory safety. These types exist to tell the compiler: "Hey, I will NOT use a 0 value, you can do whatever you want with it." As such, initializing it with 0 not only breaks the invariant but also the memory layout. For invariant correctness you should use `Result` and `.unwrap()`.

Reply:
Your understanding of how `unsafe` is designed in Rust is incomplete. `unsafe` is not restricted solely to ensuring memory safety; it also encompasses avoiding undefined behavior. Rust programs, whether they use `unsafe` code or not, **must never cause undefined behavior.** The difference is that, with `unsafe` code, the programmer assumes the responsibility of upholding certain invariants to ensure this requirement is met. 

As explicitly noted in the [Rust reference](https://doc.rust-lang.org/reference/behavior-considered-undefined.html): 

> *"Unsafe only means that avoiding undefined behavior is on the programmer; it does not change anything about the fact that Rust programs must never cause undefined behavior."*

Moreover, the reference also clarifies that **there is no formal model yet for all possible undefined behavior in unsafe Rust.** This means Rust leaves it to the programmer to carefully adhere to invariants that are necessary to prevent UB. The list of behaviors currently considered undefined is subject to change, as stated in the documentation:  

> *"This list is not exhaustive; it may grow or shrink. There is no formal model of Rust’s semantics for what is and is not allowed in unsafe code..."*

So, invariants in unsafe code are not limited to memory correctness but are crucial for avoiding any form of undefined behavior. For example, the documentation for methods like `std::slice::get_unchecked` explicitly states:

```
/// Calling this method with an out-of-bounds index is _[undefined behavior]_
/// even if the resulting reference is not used.
```

This demonstrates that the scope of invariants in `unsafe` code extends beyond just memory layout and memory access patterns—it includes anything that could lead to undefined behavior.

To apply this to the specific context of your example, the method `from_felt_unchecked` explicitly states in its documentation:

> *"This can lead to undefined behavior and big security issues."*

Thus, it is reasonable to treat this code as unsafe because violating its invariants can lead directly to undefined behavior, which is what the `unsafe` keyword is designed to guard against. Violating the invariant is not just a correctness issue—it is a potential source of UB, which makes `unsafe` absolutely appropriate here.

Reply:
In the context of `NonZero`, memory layout optimization guarantees that `Option<NonZero<T>>` has the same layout and size as `T`, with `0` representing `None`. Really cool stuff!

Reply:
![boom](https://tenor.com/vI9x.gif)

Reply:
![boom](https://media0.giphy.com/media/v1.Y2lkPTc5MGI3NjExMmNnZW1tazFpbDQyNDZtcTRtdGdtMHJ3Nnh4Zjh3bW1ndHFjYzYwZCZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/V0IdVIIW1y5d6/giphy.webp)

Reply:
While your explanation of the broader scope of `unsafe` in rust is accurate, for now, every instance of undefined behavior (UB) explicitly mentioned in the official rust documentation relates to memory safety. Rust's current model of UB, though incomplete, consistently centers around violations that affect memory layout, access, or usage in unsafe code.

The elements in the Rust reference's list suggest that the current model of `unsafe` in rust is de facto primarily focused on preserving memory safety. Even cases like using `std::slice::get_unchecked` with out-of-bounds indices fall under this umbrella, as they involve accessing memory in an unchecked manner, potentially leading to memory corruption.

The argument about invariants is valid in the theoretical scope of `unsafe`, but in practice, Rust’s usage of `unsafe` is grounded in addressing memory-related concerns.

In the case of `from_felt_unchecked`, its invariants can lead to undefined behavior due to violations impacting memory correctness, so marking it as `unsafe` is justified.

A better API in this case would be something like `from_felt(v).expect("v must not be 0")`.

Reply:
Just because most instances of undefined behavior (UB) currently documented in Rust relate to memory safety doesn't mean that the scope of `unsafe` is limited only to memory-related issues. The Rust Reference itself does not make such a statement or impose such a restriction. 

The `unsafe` keyword, as described in the Rust Reference, serves a broader purpose: it marks code that either **defines** or **satisfies** additional safety requirements. Specifically, it plays several roles depending on where it is used:

- It is used to **define** extra safety conditions in contexts such as `unsafe fn` and `unsafe trait`.
- It is used to **satisfy** extra safety conditions in contexts such as `unsafe {}`, `unsafe impl`, `unsafe extern`, or even `unsafe fn` under certain circumstances.

This perspective underscores that `unsafe` is not exclusively tied to ensuring memory safety but is instead about acknowledging and upholding additional safety invariants—whatever they may be. While memory-related issues dominate the documented cases of UB, this doesn't limit the theoretical or semantic scope of `unsafe`.

And I completely agree that a better API would involve using `Result`. In fact, such an API already exists in the form of [`TryFrom`](https://github.com/starknet-io/types-rs/blob/7ef30234da7957f7585ea832d57d13083c76995c/crates/starknet-types-core/src/felt/mod.rs#L514). So, @franciszekjob, let’s just use that, and I think we should be done here.

Reply:
Agree that it is non-exhaustive list, but as previously stated **in practice** it is memmory related. It was simplification so we don't have to make this discussion :sweat_smile:. Please stick with `TryFrom`.